Pass over Chapter 7 (#148)
pavpanchekha/emberfox: book/chrome.md

clicked on. Luckily, the event handler is passed an "event object",
whose `x` and `y` fields refer to where the click happened:

Now, here, we have to be careful with coordinate systems. Those *x*
and *y* coordinates are relative to the browser window. Since the
canvas is in the top-left corner of the window, those are also the *x*
and *y* coordinates relative to the canvas. We want the coordinates
relative to the web page, so we need to account for scrolling:

``` {.python expected=False}
class Browser:
    def click(self, e):
        x, y = e.x, e.y
        y += self.scroll
        # ...
```

The next step is to figure out what links or other elements are at
that location. To do that, we need search through the tree of layout
objects:

``` {.python indent=8}
# ...
objs = [obj for obj in tree_to_list(self.document, [])
        if obj.x <= x < obj.x + obj.width
        and obj.y <= y < obj.y + obj.height]
if not objs: return
```

Now, normally you click on lots of layout objects at once: some text,
and also the paragraph it's in, and the section that that paragraph is
in, and so on. We want the one that's "on top", so the last object in
the list:[^overlap]

> [^overlap]: In a real browser you can also have a dialog that overlaps

Reviewer: ```suggestion
[^overlap]: In a real browser you can also have sibling elements that overlap each other, such as a dialog that overlaps
```

> some text, or something like that. And real browsers can control which

Reviewer: ```suggestion
some text. And real browsers can control which
```

element is on top using the `z-index` property. So real browsers use

> *stacking contexts* to resolve what exactly you actually clicked on.

Reviewer: ```suggestion
[*stacking contexts*](https://developer.mozilla.org/en-US/docs/Web/CSS/CSS_Positioning/Understanding_z_index/The_stacking_context) to resolve what exactly you actually clicked on.
```

``` {.python indent=8}
# ...
elt = objs[-1].node
```

Now `node` refers to the specific node in the HTML tree that you
clicked on. With a link, that's usually going to be a text node. But
since we want to know the actual URL the user clicked on, we need to
climb back up the HTML tree to find the link element:

``` {.python indent=8}
# ...
while elt:
    if isinstance(elt, Text):
        pass
    elif elt.tag == "a" and "href" in elt.attributes:
        # ...
    elt = elt.parent
```

I wrote this in a kind of curious way, but this way it's easier to add

> other types of clickable things later---like text boxes and buttons in

Reviewer: ```suggestion
other types of clickable things later---like text boxes and buttons---in
```

the [next chapter](forms.md). Finally, now that we've found the link
element itself, we need to extract the URL and direct the browser to
it. This URL might be a relative URL:

``` {.python indent=12}
# ...
elif elt.tag == "a" and "href" in elt.attributes:

>     url = relative_url(elt.attributes["href"], self.url)

Reviewer: Nit: relative_url is should be called absolute_url. I'll write a PR for this.

    return self.load(url)
```

Since this needs to know the browser's current URL, we need to store
it in `load`:

``` {.python replace=Browser/Tab}
class Browser:
    def load(self, url):
        self.url = url
        # ...
```

Try it! You should now be able to click on links and navigate to new
web pages.

Multiple pages
==============

If you're anything like me, the next thing you tried after clicking on
links is middle-clicking them to open in a new tab. Every browser now
has tabbed browsing, and honestly it's a little embarrasing that our
little toy browser doesn't.

Fundamentally tabbed browsing means we need to distinguish between the
browser itself and individual tabs that browse some specific web page.
Right now the `Browser` class stores both information about the
browser (like the canvas it draws to) and information about a single
tab (like the layout tree and display list). We need to tease the two
apart.

Here's the plan: the `Browser` class will store the window and canvas,

> plus the list of tabs. Everything else goes into a new `Page` class.

Reviewer: ```suggestion
plus a list of `Tab` objects, one for each browser tab. Everything else goes into a new `Page` class.
```

Since the `Browser` stores the window and canvas, it handles all of
the events, sometimes forwarding it to the active tab.

> So the `Tab` constructor looks like this:

Reviewer: ```suggestion
The `Tab` constructor looks like this:
```

``` {.python}
class Tab:
    def __init__(self):
        with open("browser6.css") as f:
            self.default_style_sheet = CSSParser(f.read()).parse()
```

The `load`, `scrolldown`, `click`, and `draw` methods all move to
`Tab`, since that's now where all web-page-specific data lives. But
there's a change here too.

Since the `Browser` controls the canvas and handles events, it decides
when rendering happens and which tab does the drawing. After all, you

> only want one tab drawing its contents at a time! So let's remove the

Reviewer: ```suggestion
only want one tab drawing its contents at a time![^windowed] So let's remove the
```

Reviewer: Plus:

[^windowed] Unless the browser implements multiple windows!

And add an exercise for this.

``` {.python}
class Tab:
    def draw(self, canvas):
        # ...
```

Now let's turn our attention to the `Browser` class. Basically, the
`Tab` is now passive, just a collection of methods you can call to
manipulate the tab. The `Browser`'s job is to call those methods. In
the `Browser` class let's create a list of tabs and an index into it
for the active tab:

``` {.python}
class Browser:
    def __init__(self):
        # ...
        self.tabs = []
        self.active_tab = None
```

As the "active party", the `Browser` needs to handle all of the
events. So let's bind the down key and the mouse click:

``` {.python}
class Browser:
    def __init__(self):
        self.window.bind("<Down>", self.handle_down)
        self.window.bind("<Button-1>", self.handle_click)
```

These methods just unpack the event and forward it to the active tab:

``` {.python replace=e.y/e.y%20-%20CHROME_PX}
class Browser:
    def handle_down(self, e):
        self.tabs[self.active_tab].scrolldown()
        self.draw()

    def handle_click(self, e):
        self.tabs[self.active_tab].click(e.x, e.y)
        self.draw()
```

You'll need to modify the arguments `Tab`'s `scrolldown` and `click`

> methods expect. Finally, that `draw` call also just forwards to the

Reviewer: Spell this out?

```

This is all basically done now, with one small oversight: we need to
actually create some tabs! Let's start with a method that creates a
new tab:

``` {.python}
class Browser:
    def load(self, url):
        new_tab = Tab()
        new_tab.load(url)
        self.active_tab = len(self.tabs)
        self.tabs.append(new_tab)
        self.draw()
```

So this works---try it, debug it---but we're still missing a user
interface for the tabs feature. We need a way for the user to switch
tabs, create new ones, and so on. Let's turn to that next.

Browser chrome
==============

Real web browsers don't just show web page contents---they've got
labels and icons and buttons.[^ohmy] This is called the browser
"chrome";[^chrome] right now creating and listing and switching tabs
is most interesting. All of this stuff is drawn by the browser to the
same canvas as the page contents, and it requires information about
the browser as a whole (like the list of all tabs), so it has to
happen in the `Browser` class.

[^ohmy]: Oh my!

[^chrome]: Yep, that predates and inspired the name of Google's Chrome
    browser.

So let's add some code to draw a set of tabs at the top of the browser

> window. Let's try to keep it simple, because this is going to require

Reviewer: ```suggestion
window (and keep it simple, because this is going to require
```

> some tedious and mildly tricky geometry. Let's have each tab be 80

Reviewer: ```suggestion
some tedious and mildly tricky geometry). Let's have each tab be 80
```

that when the user clicks on the content, we subtract `CHROME_PX` so
that the click coordinates are relative to the "page content" portion
of the browser.

When the user clicks on the browser chrome, the browser can react:

``` {.python}
if e.y < CHROME_PX:
    if 40 < e.x < 40 + 80 * len(self.tabs) and 10 <= e.y < 40:
        self.active_tab = (e.x - 40) // 80
```

In Python, the `//` operator divides two numbers and casts the result
to an integer---in this case, the index of the tab the user clicked
on. So this code switches tabs. To try it, we'll also need to handle
the button that adds new tabs:

``` {.python}
if e.y < CHROME_PX:
    # ...
    elif 10 <= e.x < 30 and 10 <= e.y < 30:
        self.load("https://browser.engineering/")
```

Now you should be able to load multiple tabs, scroll and click around
them independently, and switch tabs whenever you want.

Navigation history
==================

Now that we are navigating between pages all the time, it's easy to
get a little lost and forget what web page you're looking at.
Browsers solve this issue with an address bar that shows the URL.
Let's implement a little address bar ourselves. The URL it shows is

> the URL for the currently-active tab:

Reviewer: ```suggestion
the one for the currently-active tab:
```
